karoo_gp/base_class.py

```diff
@@ -172,63 +173,59 @@ def __init__(self, kernel='m', tree_type='r', tree_depth_base=3,
         self.fx_data_load(filename)
-        self.gen_id = 1  # set initial generation ID
-        # initialise population_a to host the first generation
-        self.population_a = ['Karoo GP by Kai Staats, Generation ' + str(self.gen_id)]
-        # initialise population_b to satisfy fx_karoo_pause()
-        self.population_b = ['placeholder']
-        # construct the first population of Trees
-        self.fx_init_construct(tree_type, tree_depth_base)
-
-        if self.kernel == 'p':  # terminate here for Play mode
-            # print the current Tree
-            self.fx_display_tree(self.tree)
-            # save this one Tree to disk
-            self.fx_data_tree_write(self.population_a, 'a')
-            sys.exit()
-
-        elif self.gen_max == 1:  # terminate here if constructing just one generation
-            # save this single population to disk
-            self.fx_data_tree_write(self.population_a, 'a')
-            self.log('\n We have constructed a single, stochastic population of '
-                     f'{self.tree_pop_max} Trees, and saved to disk')
-            sys.exit()
+        self.fx_fitness_labels_map = fx_fitness_labels_map_maker(self.class_labels)
 
-        else:
-            self.log('\n We have constructed the first, stochastic population of ',
-                     f'{self.tree_pop_max} Trees')
+        # initialise population_a to host the first generation
+        self.population = Population.generate(
+            log=self.log, pause=self.pause, error=self.error,
+            tree_type=self.tree_type, tree_depth_base=self.tree_depth_base,
+            tree_depth_max=tree_depth_max, tree_pop_max=tree_pop_max,
+            functions=self.functions, terminals=self.terminals, rng=self.rng,
+            fitness_type=self.fitness_type
+        )
 
-        ### PART 3 - evaluate first generation of Trees ###
-        self.log('\n Evaluate the first generation of Trees ...')
-        # generate expression, evaluate fitness, compare fitness
-        self.fx_fitness_gym(self.population_a)
-        # save the first generation of Trees to disk
-        self.fx_data_tree_write(self.population_a, 'a')
+        # MAY REDESIGN - moved from init
+        self.log(f'\n\t\033[32m Press \033[36m\033[1m?\033[0;0m\033[32m at any '
+                 f'\033[36m\033[1m(pause)\033[0;0m\033[32m, or '
+                 f'\033[36m\033[1mENTER\033[0;0m \033[32mto continue the run\033[0;0m',
+                 display=['i'])
+        self.pause(display=['i'])
+        self.population.evaluate(
+            log=self.log, pause=self.pause, error=self.error,
+            data_train=self.data_train, kernel=self.kernel,
+            data_train_rows=self.data_train_rows,
+            tf_device_log=self.tf_device_log, class_labels=self.class_labels,
+            tf_device=self.tf_device, terminals=self.terminals,
+            precision=self.precision, savefile=self.savefile,
+            fx_data_tree_write=self.fx_data_tree_write
+        )
```

Comment:
Ya there is definite room for optimization. I was ignoring for now because
- The `.evaluate` method will move from `Population` to `Solver` in the next PR, so it'll be rewritten anyway
- As part of that PR, when all the main classes are there, I'll update the shared state situation, like we talked about